### PR TITLE
Fix PyTorch 1.9 backend initialization

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary

Fix the production channels-last backend owner on PyTorch 1.9, where `Tensor.is_inference()` does not exist.

- Gate inference-tensor normalization with the existing PyTorch 1.10 version constant
- Align the existing real-path assertion with the same API floor
- Preserve automatic channels-last selection, backend ownership, and PyTorch 1.8 behavior unchanged

This fixes the 117 cascading prediction, validation, training, CLI, and solutions failures exposed by the expanded slow matrix on Ubuntu/Python 3.9/PyTorch 1.9.0.

Failure: https://github.com/ultralytics/ultralytics/actions/runs/33321337963/job/99283730533

## Validation

- `test_autobackend_memory_format` passes
- `test_model_forward` passes
- Ruff and `git diff --check` pass
- Expanded PyTorch-version CI will validate the real 1.9 and 1.10 environments

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

Updated PyTorch version gating and inference-mode handling so `AutoBackend` no longer calls `Tensor.is_inference()` on PyTorch 1.9.

### 📊 Key Changes

- `AutoBackend` now normalizes inference tensors only when the existing `TORCH_1_10` version flag is enabled.
- `smart_inference_mode(False)` uses an explicit wrapper for PyTorch 1.9 to combine disabled inference mode with `torch.no_grad()`.
- `test_autobackend_memory_format` only checks `Tensor.is_inference()` on PyTorch 1.10 and newer.
- Updated an exporter warning to use a generic message and corrected the `ExecuTorch` spelling in predictor documentation comments.

### 🎯 Purpose & Impact

- PyTorch 1.9 backend initialization no longer reaches the unavailable `Tensor.is_inference()` API, while the PyTorch 1.8 path continues to use its existing behavior.
- Automatic channels-last selection and backend ownership logic remain unchanged; the test now matches the supported API floor.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The production guard in `AutoBackend` correctly avoids calling `Tensor.is_inference()` before PyTorch 1.10 and preserves the existing channels-last behavior. However, the updated test skips the affected PyTorch 1.9 initialization path, leaving the reported regression insufficiently covered.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `tests/test_python.py:225` On PyTorch 1.9, `torch.inference_mode()` exists even though `Tensor.is_inference()` does not, so this condition now skips the only test path that constructs an inference-tensor model on the affected version. The regression can therefore reappear without this test detecting it. Keep the setup/backend initialization guarded by `hasattr(torch, "inference_mode")` and guard only the `is_inference()` assertion with `TORCH_1_10`.
</details>

**Comment by glenn-jocher** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-08-30 16:25:17 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/33322355573): `ci.yml` run 30864
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/33322356410): `docker.yml` run 3120

**Review by UltralyticsAssistant — DISMISSED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed the full current diff. The prior review's test-coverage issue is addressed: the PyTorch 1.9 inference-mode setup still runs, while only the unavailable `is_inference()` assertion is gated at PyTorch 1.10. The AutoBackend guard now matches the API floor and preserves existing channels-last behavior. LGTM.

**Comment by glenn-jocher** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-08-30 16:43:23 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/33323206661): `ci.yml` run 30874
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/33323207659): `docker.yml` run 3126

File: ultralytics/utils/torch_utils.py [hunk 1]
```diff
@@ -87,11 +87,11 @@ def smart_inference_mode(mode=True):
     def decorate(fn):
         """Apply appropriate torch decorator for inference mode based on torch version."""
         if not mode:
-            return torch.inference_mode(False)(torch.no_grad()(fn)) if TORCH_1_9 else torch.no_grad()(fn)
+            return torch.inference_mode(False)(torch.no_grad()(fn)) if TORCH_1_11 else torch.no_grad()(fn)
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/torch_utils.py]: 💡 **MEDIUM**: This unrelated threshold change regresses supported PyTorch 1.9 and 1.10 behavior. Both versions provide `torch.inference_mode(False)`, but `TORCH_1_11` makes `smart_inference_mode(False)` fall back to `no_grad`; when called from an outer inference-mode context, it no longer exits inference mode. This affects `AutoBackend.__init__` and model setup/fusion paths, allowing inference tensors to be retained despite the call-site invariant that weights must remain mutable. Please preserve the existing `TORCH_1_9` guard for the `mode=False` branch and `TORCH_1_10` for the enabled branch; this PR should not modify `smart_inference_mode`. ([raw.githubusercontent.com](https://raw.githubusercontent.com/pytorch/pytorch/v1.9.0/torch/autograd/grad_mode.py))

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 3

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since review 2, the AutoBackend guard and regression-test coverage remain addressed. However, the current diff newly changes `smart_inference_mode` thresholds, introducing a PyTorch 1.9/1.10 regression in backend initialization and fusion. One medium-severity finding remains.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `ultralytics/utils/torch_utils.py:90` This unrelated threshold change regresses supported PyTorch 1.9 and 1.10 behavior. Both versions provide `torch.inference_mode(False)`, but `TORCH_1_11` makes `smart_inference_mode(False)` fall back to `no_grad`; when called from an outer inference-mode context, it no longer exits inference mode. This affects `AutoBackend.__init__` and model setup/fusion paths, allowing inference tensors to be retained despite the call-site invariant that weights must remain mutable. Please preserve the existin…
</details>

File: ultralytics/utils/torch_utils.py (hunk 1 above)
Comment (glenn-jocher) [inline ultralytics/utils/torch_utils.py]: Addressed at the shared owner. Real PyTorch 1.9 and 1.10 reproduction showed `torch.inference_mode(False)` disables inference mode as a context manager, but its decorator clone re-enables it. The enabled path remains on the existing Torch 1.10 floor; only `mode=False` now uses the working context-manager form, preserving outer-context disable semantics. Verified nested retained-model construction and autograd in real Torch 1.9.0 and 1.10.0 CPU containers.

**Comment by glenn-jocher** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-08-30 16:49:52 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/33323505944): `ci.yml` run 30878
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/33323506870): `docker.yml` run 3127

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review 4

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since review 3, the `smart_inference_mode(False)` regression is addressed with an explicit context-manager wrapper, while the PyTorch 1.9 regression setup remains covered and only the unavailable `is_inference()` assertion is gated. The AutoBackend version guard and test changes are consistent with the supported API floors. LGTM.

**Comment by UltralyticsAssistant** [comment]
🎉 The [PR](https://github.com/ultralytics/ultralytics/pull/26003) has been merged—thank you, @glenn-jocher!

> “Alone we can do so little; together we can do so much.” — Helen Keller

This contribution improves PyTorch compatibility by preventing `AutoBackend` from calling the unavailable `Tensor.is_inference()` API on PyTorch 1.9, while preserving existing behavior for PyTorch 1.8 and aligning tests with the supported API floor.
